src/nginx/t/auth_asymmetrickey.t

```diff
@@ -0,0 +1,311 @@
+
+################################################################################
+# No JWT passed in.
+
+$t->run_daemon(\&bookstore, $t, $BackendPort, 'bookstore.log');
+$t->run_daemon(\&servicecontrol, $t, $ServiceControlPort, 'servicecontrol.log');
+$t->run_daemon(\&key, $t, $PubkeyPort, $es256_pubkey, 'key.log');
+
+is($t->waitforsocket("127.0.0.1:${BackendPort}"), 1, 'Bookstore socket ready.');
+is($t->waitforsocket("127.0.0.1:${ServiceControlPort}"), 1, 'Service control socket ready.');
+is($t->waitforsocket("127.0.0.1:${PubkeyPort}"), 1, 'Pubkey socket ready.');
+
+$t->run();
+
+my $response = ApiManager::http($NginxPort,<<"EOF");
+GET /shelves?key=this-is-an-api-key HTTP/1.0
+Host: localhost
+
+EOF
+
+like($response, qr/HTTP\/1\.1 401 Unauthorized/, 'Returned HTTP 401, missing creds.');
+like($response, qr/WWW-Authenticate: Bearer/, 'Returned auth challenge.');
+like($response, qr/Content-Type: application\/json/i,
+     'Missing creds returned application/json body.');
+like($response, qr/JWT validation failed: Missing or invalid credentials/i,
+     "Error body contains 'Missing or invalid credentials'.");
+
+$t->stop_daemons();
+$t->stop();
+################################################################################
+# ES256-signed jwt token is passed in "X-Goog-Iap-Jwt-Assertion" header.
+
+$t->run_daemon(\&bookstore, $t, $BackendPort, 'bookstore.log');
+$t->run_daemon(\&servicecontrol, $t, $ServiceControlPort, 'servicecontrol.log');
+$t->run_daemon(\&key, $t, $PubkeyPort, $es256_pubkey, 'key.log');
+
+is($t->waitforsocket("127.0.0.1:${BackendPort}"), 1, 'Bookstore socket ready.');
+is($t->waitforsocket("127.0.0.1:${ServiceControlPort}"), 1, 'Service control socket ready.');
+is($t->waitforsocket("127.0.0.1:${PubkeyPort}"), 1, 'Pubkey socket ready.');
+
+$t->run();
+
+$response = ApiManager::http($NginxPort,<<"EOF");
+GET /shelves?key=this-is-an-api-key HTTP/1.0
+Host: localhost
+X-Goog-Iap-Jwt-Assertion: $es256_token
+
+EOF
+
+my ($response_headers, $response_body) = split /\r\n\r\n/, $response, 2;
+like($response_headers, qr/HTTP\/1\.1 200 OK/, 'Returned HTTP 200.');
+is($response_body, <<'EOF', 'Shelves returned in the response body.');
+{ "shelves": [
+    { "name": "shelves/1", "theme": "Fiction" },
+    { "name": "shelves/2", "theme": "Fantasy" }
+  ]
+}
+EOF
+
+$t->stop_daemons();
+$t->stop();
+################################################################################
+# RS256-signed jwt token is passed in "Authorization: Bearer" header.
+
+$t->run_daemon(\&bookstore, $t, $BackendPort, 'bookstore.log');
+$t->run_daemon(\&servicecontrol, $t, $ServiceControlPort, 'servicecontrol.log');
+$t->run_daemon(\&key, $t, $PubkeyPort, $rs256_pubkey, 'key.log');
+
+is($t->waitforsocket("127.0.0.1:${BackendPort}"), 1, 'Bookstore socket ready.');
+is($t->waitforsocket("127.0.0.1:${ServiceControlPort}"), 1, 'Service control socket ready.');
+is($t->waitforsocket("127.0.0.1:${PubkeyPort}"), 1, 'Pubkey socket ready.');
+
+$t->run();
```

Comment:
Same here. not need to stop and start daemon.

Reply:
Done.